File: docs/legacy(old PACE)/C#/PaceAPISample/sample/CreateJobTransactionUsingOverrideStatusAttribute.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using efipaceservices;

namespace Pace_Web_Service_SDK.sample
{
    class CreateJobTransactionUsingOverrideStatusAttribute
    {
        public static void Run()
        {
            CreateObjectPortType createObjectHttpBinding = PaceClient.getCreateObjectHttpBinding();
            UpdateObjectPortType updateObjectHttpBinding = PaceClient.getUpdateObjectHttpBinding();
            ReadObjectPortType readObjectHttpBinding = PaceClient.getReadObjectHttpBinding();
            CloneObjectPortType cloneObjectHttpBinding = PaceClient.getCloneObjectHttpBinding();

            Console.WriteLine("Enter the job number");
            String jobNumber = Console.ReadLine();

            Console.WriteLine("Enter the Customer code");
            String customerCode = Console.ReadLine();

            Console.WriteLine("Enter the Activity Code");
            String activityCode = Console.ReadLine();

            Job job = getJobIfExistsOrCreateNew(jobNumber, customerCode, readObjectHttpBinding, createObjectHttpBinding);

            JobCost jobTransaction = null;
            try
            {
                if (null != job)
                {
                    //close the job by setting it's status
                    job.adminStatus = "C";
                    updateObjectHttpBinding.updateJob(job);
                    Console.WriteLine("Closed the job");

                    //test adding a JobTransaction to closed status transaction && when setting overrieJobStatus unpersisted attribute on trn = true
                    try
                    {
                        jobTransaction = createJobCost(activityCode, createObjectHttpBinding, getJobPart(readObjectHttpBinding, "01", job), true, true);
                        Console.WriteLine("Successfully created the job transaction :" + jobTransaction.id);
                    }
                    catch(Exception e)
                    {
                        Console.WriteLine("Setting unpersisted fields work for normal unpersisted non calculated attributes");
                    }
                    try
                    {
                        //test adding a JobTransaction to closed status transaction && when setting overrieJobStatus unpersisted attribute on trn = false
                        // We expect this to fail, that is a valid test!
                        createJobCost(activityCode, createObjectHttpBinding, getJobPart(readObjectHttpBinding, "01", job), false, false);
                    }
                    catch (Exception e)
                    {
                        Console.WriteLine("Setting unpersisted fields work for normal unpersisted non calculated attributes");
                    }
                    //reopen job for next sample use
                    job.adminStatus = "O";
                    updateObjectHttpBinding.updateJob(job);
                }

            }
            catch (Exception e)
            {
                Console.WriteLine("Failed : " + e.Message);
            }
        }

        private static JobPart getJobPart(ReadObjectPortType readObjectHttpBinding, String partString, Job job)
        {
            JobPart part = new JobPart();
            part.job = job.job;
            part.jobPart = partString;
            return readObjectHttpBinding.readJobPart(part);
        }

        private static JobCost createJobCost(String activityCode, CreateObjectPortType createObjectHttpBinding, JobPart part, Boolean shouldOverride, Boolean posted)
        {
            JobCost jobCost= new JobCost();
            jobCost.JobPartKey = part.primaryKey;
            jobCost.activityCode = activityCode;
            jobCost.overrideJobStatus = shouldOverride;
            jobCost.posted = posted;
            return createObjectHttpBinding.createJobCost(jobCost);
        }

        private static Job getJobIfExistsOrCreateNew(String jobNumber, String customerCode, ReadObjectPortType readObjectHttpBinding, CreateObjectPortType createObjectHttpBinding)
        {
            try
            {
                Job job = new Job();
                job.job = jobNumber;
                return readObjectHttpBinding.readJob(job);
            }
            catch (Exception e)
            {
                Console.WriteLine("Job: " + jobNumber + " does not exist. Creating new job");
                Job job = new Job();
                job.customer = customerCode;
                job = createObjectHttpBinding.createJob(job);
                return job;
            }
        }
    }
}

File: docs/legacy(old PACE)/java/src/main/com/pace2020/epace/sdk/sample/CreateJobTransactionUsingOverrideStatusAttribute.java
package com.pace2020.epace.sdk.sample;

import java.rmi.RemoteException;

import com.pace2020.epace.object.Job;
import com.pace2020.epace.object.JobCost;
import com.pace2020.epace.object.JobPart;
import com.pace2020.epace.sdk.util.SecuredWSDLClient;

/**
 * @author <a href="mailto:jduval@pace2020.com">jerry duval</a>
 */
public class CreateJobTransactionUsingOverrideStatusAttribute extends SecuredWSDLClient
{
    private final String m_job;
    private final String m_customerCode;

    public CreateJobTransactionUsingOverrideStatusAttribute( final String job, final String cust )
    {
        super();

        m_job = job;
        m_customerCode = cust;
    }

    public static void main( final String[] args ) throws Exception
    {
        //Check to make sure we have 2 parameters passed in
        if( args.length != 2 )
        {
            throw new Exception( "Usage: CreateJobTransactionUsingOverrideStatusAttribute <JOB> <CUSTCODE> " );
        }
        else
        {
            final CreateJobTransactionUsingOverrideStatusAttribute clazz =
                new CreateJobTransactionUsingOverrideStatusAttribute( args[0], args[1] );

            clazz.run();
        }
    }

    public JobCost run() throws RemoteException
    {
        Job job = verifyJobExists();

        JobCost action = null;

        // first check if the parameter for job is valid
        if( null != job )
        {
            System.out.println( "Job exists, attempting to update" );

            //close the job by setting it's status
            job.setAdminStatus( "C" );
            getUpdateObjectPortType().updateJob( job );

            //test adding a JobTransaction to closed status transaction && when setting overrieJobStatus unpersisted attribute on trn = true
            action =
                createJobTransaction( getJobPart( "01", job ), Boolean.TRUE );

            try
            {
                //test adding a JobTransaction to closed status transaction && when setting overrieJobStatus unpersisted attribute on trn = false
                // We expect this to fail, that is a valid test!
                createJobTransaction(
                    getJobPart( "01", job ),
                    Boolean.FALSE );
            }
            catch( RemoteException e )
            {
                System.out
                    .println( "Setting unpersisted fields work for normal unpersisted non calculated attributes" );
            }

            //reopen job for next sample use
            job.setAdminStatus( "O" );
            getUpdateObjectPortType().updateJob( job );
        }

        return action;
    }

    private JobCost createJobTransaction( final JobPart part,
                                          final Boolean override ) throws RemoteException
    {
        JobCost jobTrans = new JobCost();

        jobTrans.setJobPartKey( part.getPrimaryKey() );
        jobTrans.setActivityCode( "000" );
        jobTrans.setOverrideJobStatus( override );

        return getCreateObjectPortType().createJobCost( jobTrans );
    }

    private Job verifyJobExists() throws RemoteException
    {
        try
        {
            Job job = new Job();
            job.setJob( m_job );
            return getReadObjectPortType().readJob( job );
        }
        catch( Exception e )
        {
            System.out.println( "Job: " + m_job + " does not exist. Creating new job" );

            // Create a instance of Job, set the customer
            Job job = new Job();
            job.setCustomer( m_customerCode );
            job = getCreateObjectPortType().createJob( job );
            return job;
        }
    }
}

